File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    const n = triangle.length;
    const dp = Array.from({length: triangle.length}, 
                          (_, i) => Array(triangle[i].length).fill(0))
    
    dp[0][0] = triangle[0][0]
    
    for (let i = 1; i < n; i ++){
        for (let j = 0; j <triangle[i].length; j++){
            if (j === 0){
                dp[i][j] = dp[i - 1][j] + triangle[i][j]
            }else if(j === triangle[i].length - 1){
                dp[i][j] = dp[i - 1][j - 1] + triangle[i][j]
            }else{
                dp[i][j] = Math.max(dp[i - 1][j - 1],dp[i - 1][j]) + triangle[i][j]
            }
        }
    }
    
    let answer = 0;
    for (let val of dp[n - 1]) {
        answer = Math.max(answer, val);
    }
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    n = len(triangle)
    dp = [[0] * i for i in range(1, n + 1)]

    dp[0][0] = triangle[0][0]

    for i in range(1, n):
        for j in range(i + 1):
            if j == 0:
                dp[i][j] = dp[i - 1][j] + triangle[i][j]
            elif j == i:
                dp[i][j] = dp[i - 1][j - 1] + triangle[i][j]
            else:
                dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i][j] 
    
    max_num = max(num for num in dp[n - 1])
    return max_num